models/engine.go

```diff
@@ -0,0 +1,81 @@
+package models
+
+import "context"
+
+// Engine defines a processing unit
+type Engine interface {
+	// Start the engine with context, errors are not expected.
+	Start(ctx context.Context) error
+	// Stop the engine.
+	Stop()
+	// Done signals the engine was stopped.
+	Done() <-chan struct{}
+	// Ready signals the engine was started.
+	Ready() <-chan struct{}
+}
+
+var _ Engine = &RestartableEngine{}
+
+// RestartableEngine is an engine wrapper that tries to restart
+// the engine in case of starting errors.
+//
+// The strategy of the restarts contains simple backoff time and
+// limited number of retries that can be configured.
+type RestartableEngine struct {
```

Comment:
We don't have multiple engines at this point. We will probably in the future convert the API server to be an engine. But still I don't think embedding restart logic inside the ingestion engine is good idea, because it increases the complexity of that component. This way I have a restartable engine unit which upon receiving an error it will restart the engine. This easily tested and validated. Having that embeeded in the event processing logic gets complex and hard to test.